src/include/migraphx/op/gathernd.hpp

```diff
@@ -47,33 +48,136 @@ struct gathernd
+            int output_lens_size   = int(q) + r - k - batch_dims - 1;
+            if(output_lens_size <= 0)
+            {
+                MIGRAPHX_THROW("GATHERND: Indices too large for static data input: k=" +
+                               std::to_string(k));
+            }
+            std::vector<std::size_t> output_lens(output_lens_size);
+            std::copy(indices_lens_iter, indices_lens_iter + (q - 1), output_lens.begin());
+            if(k < r - batch_dims)
+            {
+                auto data_lens = inputs.front().lens();
+                std::copy(data_lens.begin() + batch_dims + k,
+                          data_lens.end(),
+                          output_lens.begin() + q - 1);
+            }
+            shape output_shape{inputs.front().type(), output_lens};
+            return output_shape;
         }
-        auto indices_lens_iter = inputs.back().lens().begin();
-        auto output_lens_size  = q + r - k - batch_dims - 1;
-        std::vector<std::size_t> output_lens(output_lens_size);
-        std::copy(indices_lens_iter, indices_lens_iter + (q - 1), output_lens.begin());
-        if(k < r - batch_dims)
+        else
         {
-            auto data_lens = inputs.front().lens();
-            std::copy(
-                data_lens.begin() + batch_dims + k, data_lens.end(), output_lens.begin() + q - 1);
+            // If one or both inputs are dynamic shapes, the output is dynamic
+            auto i_shape    = inputs.back();
+            auto data_shape = inputs.front();
+            size_t k;
+            if(i_shape.dynamic())
+            {
+                // the rank of the output is a function of k, so it must be fixed.
+                if(not i_shape.dyn_dims().back().is_fixed())
+                {
+                    MIGRAPHX_THROW(
+                        "GATHERND: last dimension of indices tensor must be fixed (min=max)");
+                }
+                k = i_shape.dyn_dims().back().min;
+            }
+            else
+            {
+                k = i_shape.lens().back();
+            }
+            auto r          = data_shape.ndim();
+            auto q          = i_shape.ndim();
+            int output_ndim = int(q) + r - k - batch_dims - 1;
+            if(k > r - batch_dims)
+            {
+                MIGRAPHX_THROW("GATHERND: Indices of length " + std::to_string(k) +
+                               " cannot be used to access data of rank " +
+                               std::to_string(r - batch_dims));
+            }
+            if(batch_dims >= q or batch_dims >= r)
+            {
+                MIGRAPHX_THROW("GATHERND: rank of an input cannot be less than batch_dims=" +
+                               std::to_string(batch_dims));
+            }
+            if(output_ndim <= 0)
+            {
+                MIGRAPHX_THROW("GATHERND: Indices too large for data input: k=" +
+                               std::to_string(k));
+            }
```

Comment:
Same as  previous comment, this can't happen if we already check for `k >  r - b`